File: backend/mathweaver/agents/counter_example.py
"""Counter-Example Agent: uses Z3 to verify or find counter-examples.

Handles two input modes:
1. Cayley table: direct Z3 axiom verification
2. Conjecture: uses ConjectureHandler to test claims against known structures
"""

from __future__ import annotations

import logging
from typing import Any

from ..conjecture.handler import ConjectureHandler
from ..counterexample.forge import CounterExampleForge
from ..llm.client import extract_content
from ..models.state import AgentMessage, AgentRole
from .base import AgentContext, BaseAgent

logger = logging.getLogger(__name__)


class CounterExampleAgent(BaseAgent):
    """Verifies mathematical structures using Z3 and the counter-example forge.

    This agent has Z3 as a registered tool and uses it to:
    - Verify group axioms (closure, associativity, identity, inverse)
    - Check commutativity
    - Find non-associative operations
    - Test student conjectures against known structures
    """

    def __init__(self, forge: CounterExampleForge | None = None, llm_client: Any = None) -> None:
        super().__init__(AgentRole.COUNTER_EXAMPLE, llm_client)
        self.forge = forge or CounterExampleForge(llm_client=llm_client)
        self.conjecture_handler = ConjectureHandler(forge=self.forge)
        # Register Z3 forge as a tool
        self.register_tool("z3_verify_group", self.forge.check_group_axioms)
        self.register_tool("z3_verify_associativity", self.forge.verify_associativity)
        self.register_tool("z3_check_commutativity", self.forge.check_commutativity)
        self.register_tool("test_conjecture", self.conjecture_handler.test_conjecture)

    async def run(self, ctx: AgentContext) -> AgentMessage:
        self.call_count += 1
        abstraction = ctx.prior_results.get("abstraction", {})
        perception = ctx.prior_results.get("perception", {})
        meta = abstraction.get("metadata", abstraction)

        # Check if this is a conjecture input
        perc_meta = perception.get("metadata", {})
        if perc_meta.get("is_conjecture") or perc_meta.get("input_type") == "conjecture":
            return await self._handle_conjecture(ctx, perc_meta)

        if meta.get("structure_type") == "binary_operation":
            table = meta.get("cayley_table", [])
            if not table:
                return AgentMessage(
                    role=self.role,
                    content="未检测到运算表",
                    field_updates={},
                    confidence=0.3,
                )

            # 3.1: LLM-driven tool selection
            # The agent decides which tools to use based on context
            tool_calls_made: list[dict[str, Any]] = []
            n = len(table)

            if self.llm_client is not None:
                # Ask LLM which verification tools to use
                tool_defs = [
                    {"name": "z3_verify_group", "description": "Verify all 4 group axioms"},
                    {"name": "z3_verify_associativity", "description": "Check associativity only"},
                    {"name": "z3_check_commutativity", "description": "Check Abelian property"},
                ]
                tool_resp = await self.llm_client.chat(
                    system_prompt=(
                        "面前是一张 n×n 运算表。作为一位严谨的代数学家，"
                        "判断哪些性质值得检验。\n"
                        "可选的验证手段：\n"
                        "  z3_verify_group —— 检验四条群公理\n"
                        "  z3_verify_associativity —— 单独检验结合律\n"
                        "  z3_check_commutativity —— 检验交换律\n"
                        "返回需要执行的工具名，用逗号分隔。"
                    ),
                    user_message=f"这是一张 {n}×{n} 运算表。哪些性质需要检验？",
                )

                # Parse LLM decision
                selected_tools = set()
                for tool_def in tool_defs:
                    if tool_def["name"] in extract_content(tool_resp):
                        selected_tools.add(tool_def["name"])
                if not selected_tools:
                    # LLM chose no tools — just report the table
                    selected_tools = {"z3_verify_group"}
            else:
                # Fallback: use all tools for Cayley tables
                selected_tools = {"z3_verify_group", "z3_verify_associativity", "z3_check_commutativity"}

            # Execute selected tools (3.3: via call_tool for whitelist enforcement)
            axioms_result = None
            assoc_result = None
            comm_result = None

            if "z3_verify_group" in selected_tools:
                axioms_result = self.call_tool("z3_verify_group", table)
                tool_calls_made.append({
                    "name": "z3_verify_group",
                    "result": {"is_group": not axioms_result.success},
                    "selected_by": "llm" if self.llm_client else "default",
                })

            if "z3_verify_associativity" in selected_tools:
                assoc_result = self.call_tool("z3_verify_associativity", table)
                tool_calls_made.append({
                    "name": "z3_verify_associativity",
                    "result": {"satisfied": not assoc_result.success},
                    "selected_by": "llm" if self.llm_client else "default",
                })

            if "z3_check_commutativity" in selected_tools:
                comm_result = self.call_tool("z3_check_commutativity", table)
                tool_calls_made.append({
                    "name": "z3_check_commutativity",
                    "result": {"satisfied": not comm_result.success},
                    "selected_by": "llm" if self.llm_client else "default",
                })

            # Determine result from available data
            if axioms_result:
                is_group = not axioms_result.success
            elif assoc_result:
                # If only associativity was checked, infer group status
                is_group = not assoc_result.success
            else:
                is_group = False  # No verification done

            is_abelian = is_group and comm_result and not comm_result.success

            content = self._format_result(
                axioms_result, assoc_result, comm_result, is_group, is_abelian
            )

            # Propose field updates based on verification
            field_updates: dict[str, dict[str, Any]] = {}
            if not is_group:
                field_updates["interaction"] = {
                    "consecutive_correct": ctx.four_field_state.interaction.consecutive_correct + 1,
                }
            else:
                field_updates["cognitive"] = {
                    "backtrack_count": ctx.four_field_state.cognitive.backtrack_count + 1,
                }

            return AgentMessage(
                role=self.role,
                content=content,
                field_updates=field_updates,
                tool_calls=tool_calls_made,
                confidence=0.95,
                metadata={
                    "is_group": is_group,
                    "is_abelian": is_abelian,
                    "axiom_violation": axioms_result.counter_example if axioms_result and axioms_result.success else None,
                    "assoc_violation": assoc_result.counter_example if assoc_result and assoc_result.success else None,
                    "comm_violation": comm_result.counter_example if comm_result and comm_result.success else None,
                    "z3_level": axioms_result.level.value if axioms_result else "none",
                    "tools_selected_by": "llm" if self.llm_client else "default",
                    "selected_tools": list(selected_tools),
                },
            )

        return AgentMessage(
            role=self.role,
            content="未发现可验证的形式化结构",
            field_updates={},
            confidence=0.3,
        )

    def _format_result(self, axioms, assoc, comm, is_group, is_abelian) -> str:
        if is_group and is_abelian:
            return "四条公理悉数通过，运算可交换——这是一个交换群。"
        elif is_group:
            ce = comm.counter_example if comm else "N/A"
            return f"群公理成立，但交换律被打破。反例：{ce}"
        else:
            explanation = axioms.explanation if axioms else "未执行验证"
            return f"群公理未通过。{explanation}"

    async def _handle_conjecture(self, ctx: AgentContext, perc_meta: dict) -> AgentMessage:
        """Handle student conjectures via ConjectureHandler + Z3.

        The discovery loop: student conjectures → Z3 tests → counter-example or confirmation → Socratic prompt.
        """
        raw_text = perc_meta.get("raw_text", ctx.student_input)

        # Use call_tool for whitelist enforcement (3.3)
        result = self.call_tool("test_conjecture", raw_text)

        # Build content for collaboration agent
        verdict_map = {
            "confirmed": "猜想成立",
            "refuted": "猜想被反驳",
            "undecidable": "无法判定",
        }

        content_parts = [f"{verdict_map.get(result.verdict, '未知')}：{result.explanation}"]
        if result.counter_example:
            content_parts.append(f"反例：{result.counter_example}")
        if result.socratic_prompt:
            content_parts.append(result.socratic_prompt)

        content = "\n".join(content_parts)

        # Field updates based on conjecture result
        field_updates: dict[str, dict[str, Any]] = {}
        if result.verdict == "refuted":
            field_updates["interaction"] = {
                "consecutive_correct": ctx.four_field_state.interaction.consecutive_correct,
            }
        elif result.verdict == "confirmed":
            field_updates["interaction"] = {
                "consecutive_correct": ctx.four_field_state.interaction.consecutive_correct + 1,
            }

        return AgentMessage(
            role=self.role,
            content=content,
            field_updates=field_updates,
            tool_calls=[{"name": "test_conjecture", "result": result.to_dict()}],
            confidence=0.9,
            metadata={
                "conjecture_result": result.to_dict(),
                "conjecture_verdict": result.verdict,
                "conjecture_counter_example": result.counter_example,
                "conjecture_socratic_prompt": result.socratic_prompt,
                "is_conjecture": True,
            },
        )

File: desktop/electron/backend/agents/counter_example.ts
/**
 * Counter-Example Agent: uses the brute-force forge to verify or find counter-examples.
 *
 * Ported from Python backend (backend/mathweaver/agents/counter_example.py)
 *
 * Handles two input modes:
 * 1. Cayley table: direct brute-force axiom verification
 * 2. Conjecture: uses ConjectureHandler to test claims against known structures
 */

import type { AgentContext, AgentMessage } from '../types'
import type { LLMClient } from '../llm/client'
import { BaseAgent } from './base'
import { AgentRole, createAgentMessage } from '../types'
import { CounterExampleForge, ConjectureHandler, type ForgeResult } from '../forge/forge'
import type { ConjectureResult } from '../forge/forge'

export class CounterExampleAgent extends BaseAgent {
  forge: CounterExampleForge
  conjectureHandler: ConjectureHandler

  constructor(forge: CounterExampleForge | null = null, llmClient: LLMClient | null = null) {
    super(AgentRole.COUNTER_EXAMPLE, llmClient)
    this.forge = forge ?? new CounterExampleForge(llmClient)
    this.conjectureHandler = new ConjectureHandler(this.forge)
    // Register forge tools (3.3: whitelist enforced via callTool)
    this.registerTool('z3_verify_group', (table: number[][]) => this.forge.checkGroupAxioms(table))
    this.registerTool('z3_verify_associativity', (table: number[][]) =>
      this.forge.verifyAssociativity(table),
    )
    this.registerTool('z3_check_commutativity', (table: number[][]) =>
      this.forge.checkCommutativity(table),
    )
    this.registerTool('test_conjecture', (text: string) =>
      this.conjectureHandler.testConjecture(text),
    )
  }

  async run(ctx: AgentContext): Promise<AgentMessage> {
    this.callCount += 1
    const abstraction = ctx.prior_results['abstraction'] ?? {}
    const perception = ctx.prior_results['perception'] ?? {}
    const meta = (abstraction['metadata'] ?? abstraction) as Record<string, unknown>

    // Check if this is a conjecture input
    const percMeta = (perception['metadata'] ?? {}) as Record<string, unknown>
    if (percMeta['is_conjecture'] || percMeta['input_type'] === 'conjecture') {
      return await this.handleConjecture(ctx, percMeta)
    }

    if (meta['structure_type'] === 'binary_operation') {
      const table = (meta['cayley_table'] as number[][]) ?? []
      if (!table || table.length === 0) {
        return createAgentMessage(this.role, '未检测到运算表', {
          confidence: 0.3,
        })
      }

      // 3.1: LLM-driven tool selection
      const toolCallsMade: Record<string, unknown>[] = []
      const n = table.length
      let selectedTools: Set<string>

      if (this.llmClient !== null) {
        const toolDefs = [
          { name: 'z3_verify_group', description: 'Verify all 4 group axioms' },
          { name: 'z3_verify_associativity', description: 'Check associativity only' },
          { name: 'z3_check_commutativity', description: 'Check Abelian property' },
        ]
        const toolResp = await this.llmClient.chat(
          '面前是一张 n×n 运算表。作为一位严谨的代数学家，判断哪些性质值得检验。\n' +
            '可选的验证手段：\n' +
            '  z3_verify_group —— 检验四条群公理\n' +
            '  z3_verify_associativity —— 单独检验结合律\n' +
            '  z3_check_commutativity —— 检验交换律\n' +
            '返回需要执行的工具名，用逗号分隔。',
          `这是一张 ${n}×${n} 运算表。哪些性质需要检验？`,
        )

        selectedTools = new Set<string>()
        for (const toolDef of toolDefs) {
          if (toolResp.content.includes(toolDef.name)) {
            selectedTools.add(toolDef.name)
          }
        }
        if (selectedTools.size === 0) {
          selectedTools.add('z3_verify_group')
        }
      } else {
        // Fallback: use all tools for Cayley tables
        selectedTools = new Set<string>([
          'z3_verify_group',
          'z3_verify_associativity',
          'z3_check_commutativity',
        ])
      }

      // Execute selected tools (3.3: via callTool for whitelist enforcement)
      let axiomsResult: ForgeResult | null = null
      let assocResult: ForgeResult | null = null
      let commResult: ForgeResult | null = null

      if (selectedTools.has('z3_verify_group')) {
        axiomsResult = this.callTool('z3_verify_group', table) as ForgeResult
        toolCallsMade.push({
          name: 'z3_verify_group',
          result: { is_group: !axiomsResult.success },
          selected_by: this.llmClient ? 'llm' : 'default',
        })
      }

      if (selectedTools.has('z3_verify_associativity')) {
        assocResult = this.callTool('z3_verify_associativity', table) as ForgeResult
        toolCallsMade.push({
          name: 'z3_verify_associativity',
          result: { satisfied: !assocResult.success },
          selected_by: this.llmClient ? 'llm' : 'default',
        })
      }

      if (selectedTools.has('z3_check_commutativity')) {
        commResult = this.callTool('z3_check_commutativity', table) as ForgeResult
        toolCallsMade.push({
          name: 'z3_check_commutativity',
          result: { satisfied: !commResult.success },
          selected_by: this.llmClient ? 'llm' : 'default',
        })
      }

      // Determine result from available data
      let isGroup: boolean
      if (axiomsResult) {
        isGroup = !axiomsResult.success
      } else if (assocResult) {
        isGroup = !assocResult.success
      } else {
        isGroup = false
      }

      const isAbelian = isGroup && commResult !== null && !commResult.success

      const content = this.formatResult(
        axiomsResult,
        assocResult,
        commResult,
        isGroup,
        isAbelian,
        this.ageOf(ctx),
      )

      // Propose field updates based on verification
      const fieldUpdates: Record<string, Record<string, unknown>> = {}
      if (!isGroup) {
        fieldUpdates['interaction'] = {
          consecutive_correct: ctx.four_field_state.interaction.consecutive_correct + 1,
        }
      } else {
        fieldUpdates['cognitive'] = {
          backtrack_count: ctx.four_field_state.cognitive.backtrack_count + 1,
        }
      }

      return createAgentMessage(this.role, content, {
        confidence: 0.95,
        field_updates: fieldUpdates,
        tool_calls: toolCallsMade,
        metadata: {
          is_group: isGroup,
          is_abelian: isAbelian,
          axiom_violation:
            axiomsResult && axiomsResult.success ? axiomsResult.counterExample : null,
          assoc_violation: assocResult && assocResult.success ? assocResult.counterExample : null,
          comm_violation: commResult && commResult.success ? commResult.counterExample : null,
          z3_level: axiomsResult ? axiomsResult.level : 'none',
          tools_selected_by: this.llmClient ? 'llm' : 'default',
          selected_tools: [...selectedTools],
        },
      })
    }

    return createAgentMessage(this.role, '未发现可验证的形式化结构', {
      confidence: 0.3,
    })
  }

  private formatResult(
    axioms: ForgeResult | null,
    _assoc: ForgeResult | null,
    comm: ForgeResult | null,
    isGroup: boolean,
    isAbelian: boolean,
    ageLevel: string = 'kids',
  ): string {
    // Age-adaptive phrasing for the verification verdict.
    const abelianOk: Record<string, string> = {
      kids: '四条规则全部通过，而且碰起来的顺序可以随便换——这是一个「合作无间」的魔法家族！',
      tweens: '四条公理悉数通过，运算还满足交换律——这是一个交换群。',
      teens: '四条公理悉数通过，运算可交换——这是一个交换群。',
    }
    const groupOnlyOk: Record<string, string> = {
      kids: '四条规则都通过了，但碰起来的顺序会影响结果——顺序很重要哦！反例：',
      tweens: '群公理成立，但交换律被打破（这不是交换群）。反例：',
      teens: '群公理成立，但交换律被打破。反例：',
    }
    const failed: Record<string, string> = {
      kids: '有几条规则没通过，魔法家族还不完整。',
      tweens: '群公理未通过。',
      teens: '群公理未通过。',
    }
    if (isGroup && isAbelian) {
      return abelianOk[ageLevel] ?? abelianOk['kids']
    } else if (isGroup) {
      const ce = comm ? comm.counterExample : 'N/A'
      return `${groupOnlyOk[ageLevel] ?? groupOnlyOk['kids']}${ce}`
    }
    const explanation = axioms ? axioms.explanation : '未执行验证'
    return `${failed[ageLevel] ?? failed['kids']}${explanation}`
  }

  /** Read the current student's age band from context (default kids). */
  private ageOf(ctx: AgentContext): string {
    return (ctx.metadata['age_level'] as string) ?? 'kids'
  }

  /**
   * Handle student conjectures via ConjectureHandler + brute-force.
   * The discovery loop: student conjectures -> forge tests -> counter-example or confirmation -> Socratic prompt.
   */
  private async handleConjecture(
    ctx: AgentContext,
    percMeta: Record<string, unknown>,
  ): Promise<AgentMessage> {
    const rawText = (percMeta['raw_text'] as string) ?? ctx.student_input

    // Use callTool for whitelist enforcement (3.3)
    const result = this.callTool('test_conjecture', rawText) as ConjectureResult
    const ageLevel = this.ageOf(ctx)

    const verdictMap: Record<string, string> = {
      kids: {
        confirmed: '你的发现是对的！',
        refuted: '这个猜想被推翻了——',
        undecidable: '暂时还无法判定',
      },
      tweens: {
        confirmed: '猜想成立。',
        refuted: '猜想被反驳。',
        undecidable: '无法判定。',
      },
      teens: {
        confirmed: '猜想成立。',
        refuted: '猜想被反驳。',
        undecidable: '无法判定。',
      },
    }[ageLevel] ?? {
      confirmed: '猜想成立',
      refuted: '猜想被反驳',
      undecidable: '无法判定',
    }

    const contentParts = [`${verdictMap[result.verdict] ?? '未知'}：${result.explanation}`]
    if (result.counterExample) {
      contentParts.push(`反例：${result.counterExample}`)
    }
    if (result.socraticPrompt) {
      contentParts.push(result.socraticPrompt)
    }
    const content = contentParts.join('\n')

    // Field updates based on conjecture result
    const fieldUpdates: Record<string, Record<string, unknown>> = {}
    if (result.verdict === 'refuted') {
      fieldUpdates['interaction'] = {
        consecutive_correct: ctx.four_field_state.interaction.consecutive_correct,
      }
    } else if (result.verdict === 'confirmed') {
      fieldUpdates['interaction'] = {
        consecutive_correct: ctx.four_field_state.interaction.consecutive_correct + 1,
      }
    }

    return createAgentMessage(this.role, content, {
      confidence: 0.9,
      field_updates: fieldUpdates,
      tool_calls: [{ name: 'test_conjecture', result: result.toDict() }],
      metadata: {
        conjecture_result: result.toDict(),
        conjecture_verdict: result.verdict,
        conjecture_counter_example: result.counterExample,
        conjecture_socratic_prompt: result.socraticPrompt,
        is_conjecture: true,
      },
    })
  }
}
